iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl

```diff
@@ -0,0 +1,228 @@
+}
+
+template <typename Derived>
+inline Derived& AndThen<Derived>::and_then(const and_then_callback_t& callable) & noexcept
+{
+    Derived* derivedThis = static_cast<Derived*>(this);
+
+    if (*derivedThis)
+    {
+        callable();
+    }
+
+    return *derivedThis;
+}
+
+template <typename Derived>
+inline const Derived& AndThen<Derived>::and_then(const and_then_callback_t& callable) const& noexcept
+{
+    return const_cast<const Derived&>(const_cast<AndThen<Derived>*>(this)->and_then(callable));
+}
+
+template <typename Derived>
+inline Derived&& AndThen<Derived>::and_then(const and_then_callback_t& callable) && noexcept
+{
+    return std::move(this->and_then(callable));
+}
+
+template <typename Derived>
+inline const Derived&& AndThen<Derived>::and_then(const and_then_callback_t& callable) const&& noexcept
+{
+    return std::move(const_cast<const Derived&>(const_cast<AndThen<Derived>*>(this)->and_then(callable)));
+}
+
+template <typename Derived, typename ErrorType>
+inline Derived& OrElseWithValue<Derived, ErrorType>::or_else(const or_else_callback_t& callable) & noexcept
+{
+    Derived* derivedThis = static_cast<Derived*>(this);
+
+    if (!(*derivedThis))
+    {
+        callable(derivedThis->get_error());
+    }
+
+    return *derivedThis;
+}
+
+template <typename Derived, typename ErrorType>
+inline Derived&& OrElseWithValue<Derived, ErrorType>::or_else(const or_else_callback_t& callable) && noexcept
+{
+    return std::move(this->or_else(callable));
+}
+
+template <typename Derived, typename ErrorType>
+inline const Derived&
+OrElseWithValue<Derived, ErrorType>::or_else(const const_or_else_callback_t& callable) const& noexcept
+{
+    const Derived* derivedThis = static_cast<const Derived*>(this);
+
+    if (!(*derivedThis))
+    {
+        callable(derivedThis->get_error());
+    }
+
+    return *derivedThis;
+}
+
+template <typename Derived, typename ErrorType>
+inline const Derived&&
+OrElseWithValue<Derived, ErrorType>::or_else(const const_or_else_callback_t& callable) const&& noexcept
+{
+    return std::move(this->or_else(callable));
+}
+
+template <typename Derived>
+inline Derived& OrElse<Derived>::or_else(const or_else_callback_t& callable) & noexcept
+{
+    Derived* derivedThis = static_cast<Derived*>(this);
+
+    if (!(*derivedThis))
+    {
+        callable();
+    }
+
+    return *derivedThis;
+}
+
+template <typename Derived>
+inline Derived&& OrElse<Derived>::or_else(const or_else_callback_t& callable) && noexcept
+{
+    return std::move(this->or_else(callable));
+}
+
+template <typename Derived>
+inline const Derived& OrElse<Derived>::or_else(const or_else_callback_t& callable) const& noexcept
+{
+    return const_cast<const Derived&>(const_cast<OrElse<Derived>*>(this)->or_else(callable));
+}
+
+template <typename Derived>
+inline const Derived&& OrElse<Derived>::or_else(const or_else_callback_t& callable) const&& noexcept
+{
+    return std::move(this->or_else(callable));
+}
+
```

Comment:
Could you please add
```cpp
// BEGIN and_then
...
// END and_then
```
comments for the different interfaces? This helps to distinguish where the implementation of a new interface begins and also enables code folding on many IDEs.